File: app/src/main/java/it/bitprepared/prbm/mobile/activity/UserData.java
/*   This file is part of PrbmMobile
 *
 *   PrbmMobile is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   PrbmMobile is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with PrbmMobile.  If not, see <http://www.gnu.org/licenses/>.
 */

package it.bitprepared.prbm.mobile.activity;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import android.content.Context;
import android.graphics.Bitmap;

import it.bitprepared.prbm.mobile.model.Prbm;
import it.bitprepared.prbm.mobile.model.PrbmEntity;
import it.bitprepared.prbm.mobile.model.PrbmUnit;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Class used to collect global data related to user/application
 * Realized with singleton pattern
 */
public class UserData {

    /** Singleton instance */
    private static UserData instance = null;

    /** Base URL for HTTP Calls */
    private static final String API_BASE_URL = "http://prbm.bitprepared.it/";

    /** Reference to actual prbm */
    private Prbm prbm = null;

    /** Reference to actual entity */
    private PrbmEntity entity;
    /** Reference to actual entity column */
    private int entityColumn;
    /** List of PRBMS */
    private List<Prbm> prbmList = null;

    /** Reference to actual unit */
    private PrbmUnit unit;
    /** Serilize filename */
    private static final String FILENAME = "prbms.json";

    /** HashMap of bitmpas (for fast access) */
    private HashMap<Integer, Bitmap> backBitmaps;

    /** Reference to remote Retrofit instance */
    private RemoteInterface restInterface;

    /**
     * Empty constructor
     */
    private UserData() {
        prbmList = new ArrayList<>();
        Gson gson = new GsonBuilder()
            .setLenient()
            .create();

        Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(API_BASE_URL)
            .addConverterFactory(GsonConverterFactory.create(gson))
            .build();
        restInterface = retrofit.create(RemoteInterface.class);
    }

    /**
     * Method used to return UserData instance
     * @return Reference to singleton instance
     */
    public static synchronized UserData getInstance() {
        if (null == instance) {
            instance = new UserData();
        }
        return instance;
    }

    /**
     * Returns reference to actual prbm
     * @return Global reference to prbm
     */
    public Prbm getPrbm() {
        return this.prbm;
    }

    /**
     * Store actual prbm
     * @param p Global reference to prbm
     */
    public void setPrbm(Prbm p) {
        this.prbm = p;
    }

    /**
     * Returns reference to actual unit
     * @return Global reference to a Prbm Unit
     */
    public PrbmUnit getUnit() {
        return unit;
    }

    /**
     * Store actual unit
     * @param unit Global reference to a Prbm Unit
     */
    public void setUnit(PrbmUnit unit) {
        this.unit = unit;
    }

    /**
     * Returns reference to actual entity
     * @return Global reference to actual entity
     */
    public PrbmEntity getEntity() {
        return this.entity;
    }

    /**
     * Store actual entity
     * @param entity Global reference to a Prbm Entity
     */
    public void setEntity(PrbmEntity entity) {
        this.entity = entity;
    }

    /**
     * Returns reference to actual entity column (1-4)
     * @return Global reference to actual entity column
     */
    public int getColumn() {
        return entityColumn;
    }

    /**
     * Store actual entity column (1-4)
     * @param entityColumn Global reference to an entity column
     */
    public void setColumn(int entityColumn) {
        this.entityColumn = entityColumn;
    }

    /**
     * Check if actual unit can be moved up in its column
     * @return True if actual unit can be moved up
     */
    public boolean canMoveUnitUp() {
        List<PrbmEntity> toCheck = unit.getEntitiesFromColumn(this.entityColumn);
        if (toCheck.indexOf(entity) > 0)
            return true;
        return false;
    }

    /**
     * Check if actual unit can be moved down in its column
     * @return True if actual unit can be moved down
     */
    public boolean canMoveUnitDown() {
        List<PrbmEntity> toCheck = unit.getEntitiesFromColumn(this.entityColumn);
        if (toCheck.indexOf(entity) < toCheck.size() - 1)
            return true;
        return false;
    }

    /**
     * Store a PRBM and serialize the PRBM list
     * @param context Execution context
     * @param prbm    Prbm to be saved
     */
    public void savePrbm(Context context, Prbm prbm) {
        if (!this.prbmList.contains(prbm)) this.prbmList.add(prbm);
        saveAllPrbm(context);
    }

    /**
     * Store actual PRBM and serialize the PRBM list
     * @param context Execution context
     */
    public void savePrbm(Context context) {
        savePrbm(context, this.prbm);
    }

    /**
     * Delete a PRBM and serialize the PRBM list
     * @param context Execution context
     * @param prbm    Prbm to be deleted
     */
    public void deletePrbm(Context context, Prbm prbm) {
        if (this.prbmList.contains(prbm)) this.prbmList.remove(prbm);
        saveAllPrbm(context);
    }

    /**
     * Delete actual PRBM and serialize the PRBM list
     * @param context Execution context
     */
    public void deletePrbm(Context context) {
        deletePrbm(context, this.prbm);
    }

    /**
     * Returns list of all Prbms stored
     * @return A list of Prbms currently stored
     */
    public List<Prbm> getAllPrbm() {
        return prbmList;
    }

    /**
     * Set a loaded hashmap of bitmaps. The hashmap can be used later to improve
     * performances while retrieving bitmaps.
     * @param backBitmaps An hash map of Integer (IDs) and loaded Bitmaps
     */
    public void setBackBitmaps(HashMap<Integer, Bitmap> backBitmaps) {
        this.backBitmaps = backBitmaps;
    }

    /**
     * Return a loaded bitmaps from a Resource ID
     * @param idListImage Resource ID of requested bitmap
     * @return Already loeaded bitmap
     */
    public Bitmap getBackBitmap(int idListImage) {
        if (this.backBitmaps != null)
            return this.backBitmaps.get(idListImage);
        else
            return null;
    }

    /**
     * Serialize all prbms to private file
     * @param c Execution context
     */
    protected synchronized void saveAllPrbm(Context c) {
        FileOutputStream fos;
        try {
            Gson g = new Gson();
            File root = android.os.Environment.getExternalStorageDirectory();
            File dir = new File (root.getAbsolutePath() + "/PRBM");
            dir.mkdirs();
            String json;
            for (Prbm p : this.prbmList){
                json = g.toJson(p);
                File file = new File(dir, p.getTitle() + "-" + p.getAuthors() + "-" + p.getDate() + ".json");
                fos = new FileOutputStream(file);
                fos.write(json.getBytes());
                fos.close();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Deserialize all prbms from private file
     * @param c Execution context
     */
    protected synchronized boolean restorePrbms(Context c) {
        File root = android.os.Environment.getExternalStorageDirectory();
        File dir = new File (root.getAbsolutePath() + "/PRBM");
        String[] prbms = dir.list();
        if (prbms == null || prbms.length == 0){
            // Nothing to restore
            return true;
        }
        this.prbmList = new ArrayList<>();
        for (String file : prbms){
            if (!file.endsWith(".json")) continue;
            FileInputStream fis;
            try {
                fis = new FileInputStream(new File(dir, file));
                InputStreamReader isr = new InputStreamReader(fis);
                BufferedReader bufferedReader = new BufferedReader(isr);
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    sb.append(line);
                }
                String json = sb.toString();
                Gson gson = new GsonBuilder()
                    .registerTypeAdapter(PrbmEntity.class, new PrbmEntityDecoder())
                    .create();
                Type prbmType = new TypeToken<Prbm>(){}.getType();
                Prbm decoded = gson.fromJson(json, prbmType);
                this.prbmList.add(decoded);
            } catch (Exception e) {
                return false;
            }
        }
        return true;
    }

    public RemoteInterface getRestInterface() {
        return restInterface;
    }
}

File: app/src/main/java/it/bitprepared/prbm/mobile/activity/UserData.kt
package it.bitprepared.prbm.mobile.activity

import android.content.Context
import android.os.Environment
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.Strictness
import it.bitprepared.prbm.mobile.model.PrbmEntity
import it.bitprepared.prbm.mobile.model.PrbmEntityType
import it.bitprepared.prbm.mobile.model.PrbmEntityTypeHolder
import it.bitprepared.prbm.mobile.model.Prbm
import it.bitprepared.prbm.mobile.model.PrbmUnit
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File

/**
 * Class used to collect global data related to user/application
 * TODO: Move me to either Koin or Hilt
 */
object UserData {

    private var _prbmList: MutableList<Prbm> = mutableListOf()

    /** List of PRBMs loaded */
    // TODO: Move me to ViewModel
    @JvmStatic
    val prbmList: List<Prbm> get() = _prbmList

    /** Reference to the [Prbm] user is editing */
    // TODO: Move me to ViewModel
    @JvmStatic
    var prbm: Prbm? = null

    /** Flag to check if user is editing a PRBM */
    var editPrbm: Boolean = false

    /** Flag to check if user is editing a PRBM Unit */
    var editEntity: Boolean = false

    /** Reference to the [PrbmEntity] user is editing  */
    // TODO: Move me to ViewModel
    @JvmStatic
    var entity: PrbmEntity? = null

    private val _entityTypes: MutableList<PrbmEntityType> = mutableListOf()
    fun getEntityTypes(context: Context) : List<PrbmEntityType> {
        if (_entityTypes.isEmpty()) {
            context.resources.assets.open("prbm-entity-schema.json").use {
                _entityTypes.addAll(gson.fromJson(it.reader(), PrbmEntityTypeHolder::class.java).entities)
            }
        }
        return _entityTypes.toList()
    }
    fun getEntityTypes() = _entityTypes.toList()

    fun newEntityFromMenuIndex(context: Context, position: Int) = PrbmEntity(getEntityTypes(context)[position])

    fun getSingleChoiceEntityType(context: Context): Array<String> = getEntityTypes(context).map { it.name }.toTypedArray()

    /** Reference to actual unit  */
    // TODO: Move me to ViewModel
    @JvmStatic
    var unit: PrbmUnit? = null

    /** Returns reference to actual entity column (1-4) */
    @JvmStatic
    var column: Int = 0

    /** Base URL for HTTP Calls  */
    private const val API_BASE_URL = "http://prbmm.bitprepared.it/"

    val gson: Gson = GsonBuilder()
        .setStrictness(Strictness.LENIENT)
        .create()

    val restInterface: RemoteInterface by lazy {
        val retrofit = Retrofit.Builder().baseUrl(API_BASE_URL)
            .addConverterFactory(GsonConverterFactory.create(gson))
            .build()
        retrofit.create(RemoteInterface::class.java)
    }

    /**
     * Store actual PRBM and serialize the PRBM list
     */
    @JvmStatic
    fun savePrbm(context: Context, prbm: Prbm) {
        _prbmList.removeIf { it.uuid == prbm.uuid }
        _prbmList.add(prbm)
        serializePrbms(context)
    }

    /**
     * Delete actual PRBM and serialize the PRBM list
     */
    @JvmStatic
    fun deletePrbm(context: Context, prbm: Prbm) {
        if (_prbmList.contains(prbm)) _prbmList.remove(prbm)
        serializePrbms(context)
    }

    private fun isExternalStorageWritable(): Boolean {
        return Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED
    }

    private fun getPrbmDirectory(context: Context): File {
        if (!isExternalStorageWritable()) error("External storage not writable")
        val rootDirectory = context.applicationContext.getExternalFilesDir("")
        return File(rootDirectory, "prbms")
    }


    /**
     * Serialize all prbms to private files
     */
    @Synchronized
    fun serializePrbms(context: Context) {
        val prbmDirectory = getPrbmDirectory(context).apply {
            deleteRecursively()
            mkdirs()
        }
        _prbmList.onEach {
            // TODO Move me to PRBM class and unify filename
            File(prbmDirectory, "${it.uuid}.json").apply {
                createNewFile()
                writeText(gson.toJson(it))
            }
        }
    }

    /**
     * Deserialize all [Prbm] from private files
     */
    fun restorePrbms(context: Context) {
        val prbmDirectory = getPrbmDirectory(context)
        if (prbmDirectory.isDirectory && prbmDirectory.list().isNullOrEmpty()) {
            // Nothing to restore
            return
        }
        _prbmList.clear()
        prbmDirectory.walk().forEach { file ->
            if (file.isFile && file.extension == "json") {
                val prbm = gson.fromJson(file.readText(), Prbm::class.java)
                _prbmList.add(prbm)
            }
        }
    }
}
